modules/ballerina-core/src/main/java/org/ballerinalang/model/util/JSONUtils.java

```diff
@@ -587,12 +588,17 @@ public static void setArrayElement(BJSON json, long index, BJSON element) {
         }
         
         JsonNode jsonElement = element == null ? null : element.value();
+        ArrayNode arrayNode = ((ArrayNode) jsonNode);
         try {
-            if (jsonNode.size() <= index) {
-                throw BLangExceptionHelper.getRuntimeException(RuntimeErrors.ARRAY_INDEX_OUT_OF_RANGE,
-                    index, jsonNode.size());
+            if (arrayNode.size() <= index) {
```

Comment:
Here the array implementation is at the Jackson level. Therefore we don't have the control over it and the pointers (such as 'size', etc).
So, if we grow the array at a rate of 1.5, it will cause the array to have null values at the tail.